plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+the re-positions of lets are applied "locally" (versus globally in our case, more on that later).
+
+The first-pass remains more or less the same.
+
+The second pass is heavily altered. The original algorithm's second pass applies "local transformations": when it stumbles upon a let, it looks up its rank
+and "locally" decides if it is worth to float the let-rhs (outwards based on its rank or even inwards). It stops re-positioning / floating outwards the let-rhs,
+when it crosses over a lambda with depth one larger than its rank (level number).
+The advantage is that the lets will be floated exactly around the 'expr' that needs them, and *exactly outside a lambda definition*.
+The downside is that there is a single merged let-group geneerated in every rank. This is no problem for Haskell/GHC, because lets are letrecs.
+
+We use a different approach where we prior clean the AST PlutusIR term from all the lets (rec&nonrec) and store them in a separate 'RhsTable'.
+We use the information obtained by the dependency graph to create let-(rec&nonrec) groups at that rank. A rank is a lambda depth plus a lambda name. During
+the second pass, whenever we reach a lambda/Lambda definition we float all lets beloning to this lambda (rank, i.e. lambda's depth and name).
+This means that the lets are floated *exactly inside a lambda definition, i.e at the lambda body*, with the downside being that they are floated "a bit" more outside than what would
+be optimal as in the original algorithm (floated exactly around the 'expr' that needs them, and *exactly outside a lambda definition*).
+The upside is that this approach is not local/ad-hoc, but more holistic.
+
+-}
+
+
+-- | During the first-pass we compute a maximum "rank" for every let-binding declared in the given PIR program.
+-- See Note: [Float first-pass]
+-- A rank points to either a lambda/Lambda location or Top if the let does not depend on any lambda/Lambda.
+-- A "maximum rank" is the 'Rank' with the largest depth, i.e. the deepest lambda-dependency.
+-- A rank can also be used as a pointer to a (lambda) location in the PIR program.
+data Rank = Top -- ^ signifies that a let has no lambda/Lambda dependency and thus can be placed at the toplevel of the program
+          | LamDep Int PLC.Unique -- ^ a let has a lambda/Lambda dependency with lamDepth :: 'Int', lamName :: 'PLC.Unique'
+          deriving Eq
+
+-- | The underlying classic getter function for depth
+--
+-- Note: Top is arbitrarily defined as having depth 0.
+-- It could also be made 'minBound' or 'Word', but this is conceptually clearer and allows us to use 'IM.IntMap's.
+_depth :: Rank -> Int
+_depth = \case
+  Top -> 0
+  LamDep d _ -> d
+
+-- | Lens-style getter of the depth of a rank
+depth :: Getting r Rank Int
+depth  = to _depth
+
+instance Ord Rank where
+  compare = compare `on` _depth
```

Comment:
This doesn't satisfy the `compare x y == EQ => x == y` law, does it?

Reply:
I mean, when `d` in `LamDep` is `0`, which I guess is not supposed to happen? Worth having a comment with this invariant written out explicitly.

Reply:
Yes, `LamDep 0` should not happen. I will add a comment. The only reason we use an Int is for efficient `IntMap`